ne/engine.go

```diff
@@ -0,0 +1,229 @@
+	defer e.mutex.Unlock()
+
+	e.conversations[key] = value
+}
+
+func (e *Engine) GetConversation(key string) (string, error) {
+	e.mutex.Lock()
+	defer e.mutex.Unlock()
+
+	if value, found := e.conversations[key]; found {
+		return value, nil
+	}
+	return "", errors.New("Key not found")
+}
+
+func (e *Engine) RemoveConversation(key string) {
+	e.mutex.Lock()
+	defer e.mutex.Unlock()
+
+	delete(e.conversations, key)
+}
+
+func (e *Engine) EnqueueNotification(nr Notification) {
+	e.planChan <- nr
+}
+
+func (e *Engine) CancelNotification(uuid string) {
+	e.stopChan <- uuid
+}
+
+func (e *Engine) startNotification(nr Notification) {
+	e.mutex.Lock()
+	defer e.mutex.Unlock()
+
+	id := nr.ID()
+	// Create a new stopper channel for this plan
+	e.stoppers[id] = make(chan struct{}, 10)
+	// Store the message in the message map
+	e.notifications[id] = nr
+
+	// asyncronously launch the notification plan processing
+	go e.notificationHandler(id)
+}
+
+func (e *Engine) IsNotification(uuid string) bool {
+	e.mutex.Lock()
+	defer e.mutex.Unlock()
+
+	_, found := e.stoppers[uuid]
+	return found
+}
+
+func (e *Engine) stopNotification(uuid string) {
+	e.mutex.Lock()
+	defer e.mutex.Unlock()
+
+	// check if there are notifications going for this UUID
+	stopper, found := e.stoppers[uuid]
+	if !found {
+		return
+	}
+	log.Println("[", uuid, "]", "Sending a stop notification to the plan processor")
+
+	// It's in progress, so we'll send a message on its Stopper to
+	// be received by the goroutine executing the plan
+	stopper <- struct{}{}
+}
+
+// notificationHandler receives notification requests from the engine and steps through the plan, making
+// phone calls, sendins SMS, email, etc. as necessary
+func (e *Engine) notificationHandler(uuid string) {
+	var timerChan <-chan time.Time
+	var tickerChan <-chan time.Time
+
+	// look up notification
+	e.mutex.Lock()
+	n, found := e.notifications[uuid]
+	e.mutex.Unlock()
+	if !found {
+		return
+	}
+	// look up stop chan
+	e.mutex.Lock()
+	sc, found := e.stoppers[uuid]
+	e.mutex.Unlock()
+	if !found {
+		return
+	}
+
+	log.Println("[", uuid, "]", "Initiating notification plan")
+
+	// Iterate through each step of the plan
+	for {
+		s, lastStep := n.NextStep()
+
+		u, err := url.Parse(s.NotifyMethod())
+		if err != nil {
+			log.Println("Error parsing URI:", err)
+			log.Println("Advancing to next step in plan.")
+			continue
+		}
+		log.Println("[", uuid, "]", "Method:", s.NotifyMethod())
+
+		// This outer loop repeats a notification until it's acknowledged.  It can be broken by the expiration of the timer for this step,
+		// or by a stop request.
+	stepLoop:
+		for {
+			// Take the appropriate action, depending on the type of notification
+			switch u.Scheme {
+			case "phone":
+				e.MakePhoneCall(u.Host, n.Message(), uuid)
+			case "sms":
+				e.SendSMS(u.Host, n.Message(), uuid, false)
+			case "email":
+				e.SendEmail(fmt.Sprint(u.User, "@", u.Host), n.Message(), uuid)
+			}
+			if lastStep {
```

Comment:
You have a good point there. Either we abstract the differences away using the Notification/NotificationStep interfaces or we use two different engines here.

I was hoping to be able to put it all behind the NextStep() call of the Notification interface and keep the notification engine as its is for People as well as Teams. But you may be right that this does not work out in the end.